### PR TITLE
add description feature to test tables

### LINKED ISSUE #277: Distinguish feature/assessment reason from name
I'm working on porting Crossplane's nascent E2E tests to e2e-framework in https://github.com/crossplane/crossplane/pull/4187. I find the `features.Table` approach more readable than the builder pattern, so most of my tests look like this:

```go
func TestConfiguration(t *testing.T) {
	// Test that we can install a Configuration from a private repository using
	// a package pull secret.
	manifests := "test/e2e/manifests/pkg/configuration/private"
	private := features.Table{
		{
			Name: "ConfigurationIsCreated",
			Assessment: funcs.AllOf(
				funcs.ApplyResources(FieldManager, manifests, "*.yaml"),
				funcs.ResourcesCreatedWithin(1*time.Minute, manifests, "*.yaml"),
			),
		},
		// etc...
	}

	setup := funcs.ReadyToTestWithin(1*time.Minute, namespace)
	environment.Test(t,
		private.Build("PullFromPrivateRegistry").
			WithLabel("area", "pkg").
			WithLabel("size", "small").
			Setup(setup).Feature(),
	)
```

I've noticed that a lot of folks like to use descriptive prose to "name" their assessments and features. So more like:

```go
func TestConfiguration(t *testing.T) {
	manifests := "test/e2e/manifests/pkg/configuration/private"
	private := features.Table{
		{
			Name: "The supplied Configuration should be created",
			Assessment: funcs.AllOf(
				funcs.ApplyResources(FieldManager, manifests, "*.yaml"),
				funcs.ResourcesCreatedWithin(1*time.Minute, manifests, "*.yaml"),
			),
		},
		// etc...
	}

	setup := funcs.ReadyToTestWithin(1*time.Minute, namespace)
	environment.Test(t,
		private.Build("Test that we can install a Configuration from a private repository using a package pull secret.").
			WithLabel("area", "pkg").
			WithLabel("size", "small").
			Setup(setup).Feature(),
	)
```

I like this in theory - I think it's a good idea to provide as much written context as possible for anyone reading the test code and/or debugging why the test has failed. In practice, I stick with `PascalCase` assessment and feature names, because they're actually `t.Run` subtest names. Go mangles them into `Really_really_long/nested_snake_case_names` which I find ugly and hard to read:

```
--- PASS: TestConfiguration (60.15s)
    --- PASS: TestConfiguration/Pull_a_Configuration_from_a_private_registry (30.07s)
        --- PASS: TestConfiguration/Pull_a_Configuration_from_a_private_registry/The_supplied_Configuration_should_be_created (5.03s)
        --- PASS: TestConfiguration/Pull_a_Configuration_from_a_private_registry/The_supplied_configuration_should_become_healthy,_and_active (5.01s)
        --- PASS: TestConfiguration/Pull_a_Configuration_from_a_private_registry/The_supplied_Configuration_should_be_deleted (5.03s)
    --- PASS: TestConfiguration/Pull_a_Configuration_that_depends_on_a_Provider (30.08s)
        --- PASS: TestConfiguration/Pull_a_Configuration_that_depends_on_a_Provider/The_supplied_Configuration_should_be_created (5.02s)
        --- PASS: TestConfiguration/Pull_a_Configuration_that_depends_on_a_Provider/The_supplied_Configuration_should_become_healthy,_and_active (5.01s)
        --- PASS: TestConfiguration/Pull_a_Configuration_that_depends_on_a_Provider/The_Provider_the_Configuration_depends_on_should_become_healthy,_and_active (5.01s)
        --- PASS: TestConfiguration/Pull_a_Configuration_that_depends_on_a_Provider/The_supplied_Configuration_should_be_deleted (5.01s)
        --- PASS: TestConfiguration/Pull_a_Configuration_that_depends_on_a_Provider/The_supplied_Provider_should_be_deleted_(providers_are_not_deleted_automatically) (5.01s)
```

What do you think about allowing folks to provide a feature or assessment's written context, distinct from the _name_ of the feature or assessment? For example:

```go
func TestConfiguration(t *testing.T) {
	manifests := "test/e2e/manifests/pkg/configuration/private"
	private := features.Table{
		{
			Name: "CreateConfiguration",
			Reason: "We should be able to install a Configuration from a private repository using a package pull secret",
			Assessment: funcs.AllOf(
				funcs.ApplyResources(FieldManager, manifests, "*.yaml"),
				funcs.ResourcesCreatedWithin(1*time.Minute, manifests, "*.yaml"),
			),
		},
		// etc...
	}

	setup := funcs.ReadyToTestWithin(1*time.Minute, namespace)
	environment.Test(t,
		private.Build("PullFromPrivateRegistry").
			Reason("Crossplane should support pulling packages from private repositories").
			WithLabel("area", "pkg").
			WithLabel("size", "small").
			Setup(setup).Feature(),
	)
```

I imagine the `Reason` field would be emitted via `t.Log` when a test started, and perhaps somehow included (regardless of test verbosity) if a test failed. I first saw a pattern like this [in the go-cmp unit tests](https://github.com/google/go-cmp/blob/v0.5.9/cmp/compare_test.go#L176), and we adopted it a while back for Crossplane unit tests to good effect.

### PR DESCRIPTION
**Pull request by harshanarayana**
Enable Test Table Description capabilities

Closes #277

### CODE REVIEW

**Comment by harshanarayana** [comment]
```bash
=== RUN   TestTableDriven
=== RUN   TestTableDriven/Random_numbers
=== RUN   TestTableDriven/Random_numbers/testWithDescription
    /Users/harshnar/Work/Tesseract/Maglev/mks/k8s/e2e-framework/examples/table/table_test.go:52: Looks like the current time could not be determined
=== RUN   TestTableDriven/Random_numbers/less_than_equal_64
    /Users/harshnar/Work/Tesseract/Maglev/mks/k8s/e2e-framework/examples/table/table_test.go:63: limit should be less than 64
=== RUN   TestTableDriven/Random_numbers/more_than_than_equal_128
    /Users/harshnar/Work/Tesseract/Maglev/mks/k8s/e2e-framework/examples/table/table_test.go:74: limit should be less than 128
=== RUN   TestTableDriven/Random_numbers/Assessment-3
    /Users/harshnar/Work/Tesseract/Maglev/mks/k8s/e2e-framework/examples/table/env.go:450: Failed Executing Test: This is an example of how to create a test with description
=== RUN   TestTableDriven/Feature-2
=== RUN   TestTableDriven/Feature-2/A_simple_feature
    /Users/harshnar/Work/Tesseract/Maglev/mks/k8s/e2e-framework/examples/table/table_test.go:98: this is a great number
--- FAIL: TestTableDriven (0.00s)
    --- FAIL: TestTableDriven/Random_numbers (0.00s)
        --- FAIL: TestTableDriven/Random_numbers/testWithDescription (0.00s)
        --- PASS: TestTableDriven/Random_numbers/less_than_equal_64 (0.00s)
        --- PASS: TestTableDriven/Random_numbers/more_than_than_equal_128 (0.00s)
        --- PASS: TestTableDriven/Random_numbers/Assessment-3 (0.00s)
    --- PASS: TestTableDriven/Feature-2 (0.00s)
        --- PASS: TestTableDriven/Feature-2/A_simple_feature (0.00s)
FAIL
FAIL	sigs.k8s.io/e2e-framework/examples/table	1.218s
```

Example of test description being logged as part of the failure.

File: pkg/env/env.go [hunk 1]
```diff
@@ -442,6 +442,16 @@ func (e *testEnv) execFeature(ctx context.Context, t *testing.T, featName string
 		failed := false
 		for i, assess := range assessments {
 			assessName := assess.Name()
+			if dAssess, ok := assess.(types.DescribableStep); ok {
+				if dAssess.Description() != "" {
+					klog.V(4).InfoS("Processing Assessment", "description", dAssess.Description())
+					defer func() {
+						if t.Failed() {
+							t.Logf("Failed Executing Test: %s", dAssess.Description())
+						}
```
Comment (negz) [inline pkg/env/env.go:451]:
```suggestion
					t.Logf("Assessment description: %s", dAssess.Description())
```

Would this work? My goal is that:

* The description is printed in `-test.v` mode even if the test _doesn't_ fail.
* The description is printed (with or without `-test.v`) if the test does fail.

I don't think we need to explicitly state that the test failed - it's fine for the reader to see (roughly):

* TestFoo: Assessment description: ensure that something something...
* TestFoo: FAILED

File: pkg/features/builder.go [hunk 2]
```diff
@@ -71,6 +76,10 @@ func (b *FeatureBuilder) Assess(desc string, fn Func) *FeatureBuilder {
 	return b.WithStep(desc, types.LevelAssess, fn)
 }
 
+func (b *FeatureBuilder) AssessWithDescription(name, description string, fn Func) *FeatureBuilder {
+	return b.WithStepDescription(name, description, types.LevelAssess, fn)
+}
```
Comment (negz) [inline pkg/features/builder.go:81]: It would be useful for the whole feature (not only each assessment) to have a description too.

**Review by negz — COMMENTED** [review]
Thanks for tackling this @harshanarayana! I left some thoughts.

File: pkg/env/env.go (hunk 1 above)
Comment (harshanarayana) [reply pkg/env/env.go:451]: Makes sense. Cleaned it up. What you suggested is much easier.

**Review by vladimirvivien — COMMENTED** [review]
This is great @harshanarayana.
I think this is enough for now. If someone asks for filtering based on description that can be added later.

**Comment by harshanarayana** [comment]
@vladimirvivien  If you are good with these changes, we can get this one merged?

**Pull request merged**